### ISSUE TITLE
Command-line interface for supervise

### ISSUE THREAD

**Issue by tailhook** [tailhook/vagga#338]
# Motivation

Since we now run a lot of test suites using `!Supervise` command we really want to be able to pass some params via the command-line to tests.

While it's currently possible via `vagga -E X=Y tests` it's not very friendly. Also, the great point of vagga is introspection of commands (i.e. you can run `vagga` or `vagga _list`)

So we definitely want some thing, that will allow `vagga tests --help` to work, and also:

1. Keeps `--exclude` and `--only` working too
2. Allows to pass different parameters to different processes (in the same run)

# Proposal

My current idea is to do something like this:

```yaml
commands:
  tests: !Supervise
    description: Run test suite
    options: |
      Usage:
        vagga test <test_name>...

      Options:
        -h --help     Show this screen.
        --redis-port <num>
                      Port for redis server [default: 6379]
    children:
      tests: !Command
        container: py
        run: |
          py.test --redis-port=$VAGGAOPT_REDIS_PORT $VAGGAOPT_TEST_NAME
      redis: !Command
        container: redis
        run: |
          redis-server --daemonize no --port $VAGGAOPT_REDIS_PORT
```
Basically, it means using [docopt](https://github.com/docopt/docopt.rs) for parsing the options passing them using environment.

# Questions

1. Does it worth it?
2. Isn't `VAGGAOPT_` too long? Should we use `VO_`, `OPT_` or no prefix at all?
3. Should description and built-in options be appended automatically or should they be explicitly listed here (with some convention)
4. What's the policy on adding a new built-in option? The issue is that any option can conflict with user config.
5. Should we have a "pass all other options to a command" thing? Is it possible with docopt?
6. How to pass multiple arguments in a single variable (vector or list of values)? I.e. how arguments should be escaped.

# Alternatives

1. Pass all excessive arguments to a command as single env variable, like `CMD_ARGS`. *This means you need to parse command-line in each of commands just to filter out irrelevant options.*
2. Use some Yaml'y syntax for command-line args. *This probably means more learning. And I think docopt will be more readable, although less functional theoretically*
3. Add params/substitute vars by vagga instead of using shell and env variables. *This introduces additional layer of variables in vagga config, so requires more learning.*

/cc @popravich , @anti-social , @lilianmoraru , @livioribeiro

**Comment by livioribeiro** [comment]
1. It is very interesting, probably worth it.
2. I think we don't need a prefix here since the user knows the environment and can avoid collisions.
3. Options could be appended to Description, but Description has a chance to become obsolete.
4. Maybe a warning of the conflicting option before adding the built-in one?
5. It could be useful for commands that, for example, receive a list of files to process.
6. There could be a repetition syntax, for example:
```yaml
options: |
    Usage:
      vagga process <files>...
run: process {{ -f FILES }} # or: process -f {{ FILES }}
```
The `run` command would expand to `process -f $FILES_0 -f $FILES_1 -f $FILES_2` (or `process -f $FILES_0 $FILES_1 $FILES_2`).

The `options` could also be added to regular commands to provide validation before executing the container.

**Comment by tailhook** [comment]
> I think we don't need a prefix here since the user knows the environment and can avoid collisions.

Well, there are a lot of generic environment variables like `PATH` and `LANG`. And we don't have a way to rename an option in docopt AFAIU, i.e. either we can't name arguments `--path`, `--lang` or we have to add a prefix to env var names.

**Comment by livioribeiro** [comment]
Oh, I totally forgot about that

**Comment by tailhook** [comment]
Okay, initial implementation is in #338 if anybody wants to review.

**Comment by popravich** [comment]
My only consern is that it may be hard to write `options` in this format.
Also it may be painful to deal with merge conflicts in suct texts.

**Comment by tailhook** [comment]
> My only consern is that it may be hard to write options in this format.
> Also it may be painful to deal with merge conflicts in suct texts.

Any alternatives?

**Comment by popravich** [comment]
> Any alternatives?

No idea yet. Using Yaml structures doesn't sound good either

**Comment by tailhook** [comment]
Well, I don't think it's hard. Basically, it's like writing a help string. There is a pretty good description at [docopt.org](http://docopt.org) and there is [try.docopt.org](http://try.docopt.org)

Merging shouldn't be a problem too. The format is basically line-based. I.e. every option is on the new line.

I'm more in doubt of how to write bash scripts that use options/env vars. Some scenarios are pretty hard to express in bash. And even harder in ash or dash (alpine and ubuntu default shell respectively).

**Comment by tailhook** [comment]
Merged #338 in. It will be available in the testing repository shortly. Please try and report any issues!
